lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp

```diff
@@ -515,6 +515,54 @@ void TypeLoweringVisitor::visitDecl(MemOp op) {
   opsToRemove.push_back(op);
 }
 
+/// Copy annotations from \p annotations to \p loweredAttrs, except annotations
+/// with "target" key, that do not match the field suffix.
+static void filterAnnotations(ArrayAttr annotations,
+                              SmallVector<Attribute> &loweredAttrs,
+                              const FlatBundleFieldEntry &field,
+                              MLIRContext *context) {
+
+  for (auto opAttr : annotations) {
+    auto di = opAttr.dyn_cast<DictionaryAttr>();
+    if (!di) {
+      loweredAttrs.push_back(opAttr);
+      continue;
+    }
+    auto targetAttr = di.get("target");
+    if (!targetAttr) {
+      loweredAttrs.push_back(opAttr);
+      continue;
+    }
+
+    ArrayAttr subFieldTarget = targetAttr.cast<ArrayAttr>();
+    SmallString<16> targetStr;
+    for (auto fName : subFieldTarget) {
+      std::string fNameStr = fName.cast<StringAttr>().getValue().str();
+      // The fNameStr will begin with either '[' or '.', replace it with an
+      // '_' to construct the suffix.
+      fNameStr[0] = '_';
+      // If it ends with ']', then just remove it.
+      if (fNameStr.back() == ']')
+        fNameStr.erase(fNameStr.size() - 1);
+
+      targetStr += fNameStr;
+    }
+    if (!targetStr.empty()) {
+      auto pos = field.suffix.find(targetStr.str().str());
+      if (pos == 0) {
+        NamedAttrList modAttr;
+        for (auto attr : di.getValue()) {
+          if (attr.first.str() == "target")
+            continue;
+          modAttr.push_back(attr);
+        }
+        loweredAttrs.push_back(DictionaryAttr::get(context, modAttr));
+      }
+    } else
+      loweredAttrs.push_back(opAttr);
```

Comment:
Check my logic here, but the nesting can be reduced a fair amount. (Comments would help with this structure, too):

```suggestion
    if (targetStr.empty()) {
      loweredAttrs.push_back(opAttr);
      continue;
    }

    if (field.suffix.find(targetStr.str().str()) != 0)
      continue;

    NamedAttrList modAttr;
    for (auto attr : di.getValue()) {
      if (attr.first.str() == "target")
        continue;
      modAttr.push_back(attr);
    }

    loweredAttrs.push_back(DictionaryAttr::get(context, modAttr));
```